### DISCUSSION TITLE
How to execute the step definition/ feature (test-case) with xunit framework

### DISCUSSION THREAD

**Discussion by pritesh848** [reqnroll/Reqnroll#261] [Q&A] [answered]
How to execute the step definition (test-case) with xunit framework, if I use [FACT] attribute it was executing every method in step-definition file as a testcase and I want to run testcase with combination of all methods in step definition  complete feature file.

**Comment by UL-ChrisGlew** [comment]
Just so I can understand better - are you using the BDD (.feature) files for your test cases or are we talking about unit testing in .cs files in your project?

**Reply by pritesh848** [reply]
Yes I am using my .feature file covered with testcase scenario and same is implemented in stepdefination file 
I am executing stepdefination class as a testcase by using [binding] attribute getting above error

 Test method ReqnrollQuickstart.Specs.Features.**********************ScreenAndVerifyUIElements threw exception: 
Reqnroll.BoDi.ObjectContainerException: Interface cannot be resolved: Xunit.Abstractions.ITestOutputHelper (resolution path: ReqnrollQuickstart.Specs.StepDefinitions.********************_AutoStepDefinitions)

**Reply by UL-ChrisGlew** [reply]
So inside your Step Definition file, you're declaring Test Cases? If so, that's not quite how Reqnroll is supposed to work. A common structure is this:

- Feature Files (these should contain all your test cases that you want to execute written in BDD/Gherkin format)
- Step Definition Files (these should contain methods decorated with only a Given/Then/When attribute).

As an example, below would contain a valid feature file:

```
Feature: My Test Feature

Scenario: I want to test something

Given I do some setup
When I test something
Then something else happens
```

An example Step Definition file would look like:

```
[Binding]
public void ExampleStepDefinitions
{
    [Given("I do some setup")]
    public void DoSomeSetup()
    {
        // Do something in here.
    }

    [When ("I test something")]
    public void ITestSomething()
    {
        // Do something in here.
    }

    [When ("something else happens")]
    public void SomethingElseHappens()
    {
        // Do something in here.
    }
}
```

You don't need to decorate each step definition with anything else other than the Given/Then/When attribute. 

The beauty of Reqnroll/SpecFlow is that they will parse the feature files to generate your unit tests for you - you should be able to see a generated `.feature.cs` file present for each `.feature` file that exists in your project.

**Reply by pritesh848** [reply]
ok thanks,  but in this case where to put mobile app capability base class, because in current framework I am using base class and extending it in testcase to launch the mobile application ... If I put it in given method then it will be add in every testcase given method ... anyway possible way for mobile automation to add app launch method in background or any different logic?

**Reply by ajeckmans** [reply]
Please take a look at https://docs.reqnroll.net/latest/automation/hooks.html And for a complex example take a look at https://github.com/reqnroll/Sample-ReqOverflow

**Reply by UL-ChrisGlew** [reply, marked as answer]
So if you wanted to perform a common setup in every test case, you should use [Reqnroll Hooks](https://docs.reqnroll.net/latest/automation/hooks.html) to perform more global setup.

So you could add a file that has something like:

```
[Binding]
public class SetupHooks
{
    [BeforeScenario]
    public void LaunchMobileApplication()
    {
        // Do my mobile setup here 
    }
}
```

this would then run the contents of the `LaunchMobileApplication` before every test case in your test suite.

**Reply by pritesh848** [reply]
Thanks @UL-ChrisGlew it helps...

How to get the report (which I can be share) 
Currently after executing an ReqmRoll bdd test, in XUNITtest explorer getting the results with girkin sentences I mentioned in feature file.
![results reqnroll](https://github.com/user-attachments/assets/28944f36-71cb-4eca-be4d-5309b3907e7a)

**Reply by gasparnagy** [reply]
For reporting you can use Allure Reports or simply run the tests from command line with `dotnet test --logger trx` and generate a HTML report from it for example with this tool: https://github.com/SamhammerAG/TrxToHtml
